rfc/20240610-Tofu-Version-Compatibility.md

```diff
@@ -0,0 +1,106 @@
+
+```hcl
+# version.tf
+terraform {
+    required_version = "1.7.4"
+}
+```
+```hcl
+# version.tofu
+tofu {
+    required_version = "1.6.5"
+}
+```
+
+When loaded by:
+* Terraform, the `version.tofu` file would be ignored and the "Terraform Required Version" would be 1.7.4.
+* OpenTofu, the `version.tofu` file would be loaded instead of `version.tf` and the "OpenTofu Required Version" would be 1.6.5.
+
+
+What happens if only the `version.tf` file exists in this module with no tofu overwrite? This is a common case for modules who have not yet considered explicit OpenTofu migration.
+
+As a user, I would like to know if any incompatible terraform_version has been specified as part of my project's configuration and modules. I would also assume any version requirement under 1.6 would be identical between Terraform and OpenTofu.
+
+```hcl
+terraform {
+    required_version = 1.7.4
+}
+```
+```shell
+$ TF_LOG=debug tofu init
+Initializing the backend...
+Initializing modules...
+Warning: Configuration requests terraform version >= 1.6!
+  Please update the configuration to specify the required OpenTofu Version.  More information is available in the debug log.
+Debug: terraform required_version = 1.7.4 in main.tf:421
+```
+
+This could also include some information on how to remedy this situation, perhaps linking to the docs.
+
+### Technical Approach
+
+The "required_version" field is accessed via `sniffCoreVersionRequirements(body)` in internal/configs/parser_config.go and stored in `configs.File.CoreVersionConstraints`. These are then merged into `configs.Module.CoreVersionConstraints` and are checked in `configs.Module.CheckCoreVersionRequirements()`. This check function is called in a variety of locations moving up the stack, but is as high as we need to concern ourselves with for this RFC. We will need to make sure the warning won't be emitted multiple times due to the spaghettified command package.
+
+We have two very similar paths here:
+* Split CoreVersionConstraints into two fields: `TofuVersionConstraints` and `TerraformVersionConstraints`
+* Add a flag into each VersionConstraint that understands if it is Tofu or Terraform specific.
+
+Either path is easy to both implement and thoroughly test.
+
+The `configs.Module.CheckCoreVersionRequirements()` function is where we will need to emit the warnings and corresponding debug log entries.
+
+### Open Questions
+
+Should there be an option (config, CLI, or ENV) to:
+* Disable the version checking altogether?
+* Don't check terraform required_versions
+* Treat both Tofu and Terraform version as identical instead of using our equivalent "terraform version" guess?
+* Do we show the guessed terraform version in `tofu -version`, maybe with a `-verbose` flag?
```

Comment:
This can be a good add on, as displaying the guessed Terraform version provides users with immediate feedback and ensures they are using the correct version for their configurations.

Reply:
@janosdebugs above reminded us that the TSC does not want a version compatibility table.   Perhaps we could link to the migration docs instead?